### PR TITLE
Removing dotted lists from cons, set-cdr, and streams

### PR DESCRIPTION
**Pull request by dnachum**
I made changes based on [this document](https://docs.google.com/document/d/1LnpnNY8hr_XmbAMvxFGuSjKjF4M2wvWO1SA1TFlKRpA/edit?usp=sharing)

Cons and set-cdr only allow cdr to be a well formed pair or nil
example_cons

Promises now, when forced, will error if it does not contain a well formed list, stream, or nil
example_promise

Because of the change to promises, streams now work like this
example-stream

~~I want to take a little bit more time to take a look at removing the dot when someone inputs something like '(1 2 3 . 4) and with variadic functions since I want to make sure I'm not breaking something unintentionally.~~

Nevermind, I made these changes but not sure if there is a better way to do it.

### CODE REVIEW

File: lib/src/core/standard_library.dart [hunk 1]
```diff
@@ -113,7 +113,13 @@ class StandardLibrary extends SchemeLibrary with _$StandardLibraryMixin {
   Value cdr(Pair val) => val.second;
 
   /// Constructs a pair from values [car] and [cdr].
-  Pair cons(Value car, Value cdr) => Pair(car, cdr);
+  Pair cons(Value car, Value cdr) {
+    // Added if statement to disallow malformed lists
+    if (!(cdr is PairOrEmpty && cdr.wellFormed)) {
```
Comment (kavigupta) [inline lib/src/core/standard_library.dart]: `(cons 1 (delay x))` should be valid

**Comment by kavigupta** [comment]
Also, we should disable quoted dotted list and implement variadic functions in the new style

**Comment by jathak** [comment]
@kavigupta Has John signed off on this change for future semesters? I'm worried that these changes may be more fragile and difficult to revert in this interpreter due to the static type system (especially with #76 wrapping `Pair` objects known to be well-formed lists in `SchemeList`).

One thing I would propose to make this less fragile would be adding different language options similar to [Racket](https://racket-lang.org/). Major changes to language semantics would add a new language, with the old semantics still available to switch back to. We can run tests with both the old and new semantics to ensure changes to unrelated parts of the interpreter don't break either.

**Comment by jathak** [comment]
I could probably create a PR off of this branch to add the basics of language options (at least internally; would take more time to add options for users) before Monday. I'll try to keep as much of this work outside the private impl repo as I can, but anything that does need to be there would have to be done by one of you  .

**Comment by jathak** [comment]
I've started work on multiple language support on the lang-options branch. It's not ready yet, but the [changes to ProjectInterface](https://github.com/Cal-CS-61A-Staff/dart_scheme/blob/lang-options/lib/src/core/project_interface.dart) should be stable if one of you wants to start it.

**Comment by kavigupta** [comment]
Hi Jen! 

The changes to berkeley-cs61a should be fairly easy to revert, as I've kept track of [them](https://github.com/Cal-CS-61A-Staff/berkeley-cs61a/issues/2388).

In addition, the staff interpreter has an option `--dots-are-cons` that should revert to older behavior, in case we need to run old code for one reason or another.

I hope that this change persists, but it is primarily intended as an experiment for this semester (Garcia likes the idea) and the summer (@chris-allsman is on board for keeping it over the summer), so revertability is key.

If it works out (students are less confused by the lack of dots and still understand the pair concept), it will remain in the course.

**Comment by dnachum** [comment]
Thanks for adding the language options Jen! I was able to make changes to the private impl repo files to get it to work so should I maybe make another PR?

**Comment by jathak** [comment]
Yeah. We'll also need to update Travis to use the new private impl. I'll send you instructions for that via Facebook

**Comment by dnachum** [comment]
Alright, thanks! I pushed the changes!

**Comment by jathak** [comment]
Closed in favor of #89

**Pull request closed without merging**